Use base images from "armhf" org for consoles & resizefs

djmaze opened the pull request:
This is a temporary measure to get the consoles working on ARM until Docker supplies real multi-arch base images.

I can understand anyone who considers this a hack. So feel free not to accept this PR ;)

luxas commented on src/docker/10-debianconsole/.prebuild.sh:
```diff
@@ -1,6 +1,15 @@
+# FIXME Remove this when Docker supplies multi-arch base images
+if [ "$ARCH" == "arm" ]; then
+  BASE_IMAGE=debian:jessie
+  ARM_BASE_IMAGE=armhf/debian:jessie
```
I don't think this is working on RPi 1. Maybe use `resin/rpi-raspbian:jessie`?

djmaze replied:
I just had a look at `resin/rpi-raspbian` and it doesn't seem to be different. It's also an `armhf` image, so it will probably work on RPi1's with hard-float OS.

luxas replied:
I've tested it and `armhf/debian` doesn't work on RPi 1. `resin/rpi-raspbian` does.
I don't know why `armhf/*` images fail on RPi 1 when it's `armhf`, but I think it has something to do with `VFPv3` support or something (compare `GOARM=6` and `GOARM=7`)

ibuildthecloud replied:
For RPi 1s we have to use `armel/debian` because debians armhf is ARMv7 where as RPi 1s are ARMv6.  Now I don't think an `armel/ubuntu` exists so the Ubuntu console might never work on a RPi 1.  Probably not a big loss.

djmaze replied:
Ah, finally I understand. `resin/rpi-raspbian` builds on Raspbian, which is specially compiled [to support ARMv6 including hard-float](https://www.raspbian.org/RaspbianFAQ#What_is_Raspbian.3F) (didn't know that). For the RPi1, that would be the best solution. And it works for newer devices as well.

So we could use `resin/rpi-raspbian` for the debian console and `armhf/ubuntu` for the ubuntu one (with the latter one not working on RPi1).

luxas replied:
Yep, that's how it works. People often consider `armhf` == ARMv7 (e.g. all `armhf/*` images), even though RPi has an armhf ARMv6 processor.
`resin/rpi-raspbian` and `armhf/ubuntu` sgtm

luxas commented:
The other option is replacing the a placeholder (e.g. `BASEIMAGE`) with `sed` to a value (e.g. `DEBIAN_BASE_IMAGE`) defined in a config file. Also commented [here](https://github.com/rancher/os-images/commit/645e42d532199dca31f82a5671f36393afb2f94d#commitcomment-15784641)

We should also be able to release both `arm` and `amd64` images from an `amd64` host

ibuildthecloud commented:
@luxas I don't know if building ARM images on x86_64 is really a good long term strategy.  It seems like we can have more straight forward builds it we just assume you need an ARM machine to do an ARM build.

@djmaze I think I might prefer @luxas idea of putting some token in the Dockerfile and then replacing it with the correct base image.  That or we put in the Dockerfile something like `rancher/os-debianconsole-base` which is an image that never exists but in the prebuild you label `debian` or `armel/debian` as `rancher/os-debianconsole-base`.

luxas commented:
@ibuildthecloud Can we use Travis for this? Also it should be documented that ARM builds should be run on an ARM machine.

djmaze commented:
Doing string replacements in the Dockerfiles means changing versioned files, which is quite ugly. I instead went the second route @ibuildthecloud suggested. 
- The concerning images are now using "virtual" base images which get tagged in the respective prebuild script.
- The (architecture-specific) base images are configured in a .env file per architecture.

If this looks good to you, I can squash the changes and rebase on the latest changes.

ibuildthecloud commented on src/docker/10-debianconsole/.prebuild.sh:
```diff
@@ -1,6 +1,11 @@
 
+# FIXME Remove this when Docker supplies multi-arch base images
+. .$ARCH.base-images
+docker pull $DEBIAN_BASE_IMAGE
```
In general in our scripts we try not to pull if the image is already local, just to speed things up.  So you can you add a `docker inspect $DEBIAN_BASE_IMAGE >/dev/null 2>&1 || docker pull ...`

djmaze replied:
Ok,  changed. I had to do the check a bit differently because `docker inspect` returns an error if the image does not exist, which does not work with `set -e`.

luxas commented:
@djmaze Can you squash the commits so we can get this in?

luxas commented:
Plus adding arches for the new `10-*` images: `fedora` and `centos`

djmaze commented:
Adjusted the additional images and squashed the commits.

I didn't find an up-to-date CentOS 7 image for ARM so I built my own. Feel free to change if you know a better one.

ibuildthecloud commented:
LGTM

imikushin commented:
LGTM. Merging it
